web/packages/teleport/src/SessionRecordings/view/Timeline/renderers/FramesRenderer.test.ts

```diff
@@ -0,0 +1,197 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import 'jest-canvas-mock';
+
+import { darkTheme } from 'design/theme';
+
+import { SessionRecordingThumbnail } from 'teleport/services/recordings';
+import type { TimelineRenderContext } from 'teleport/SessionRecordings/view/Timeline/renderers/TimelineCanvasRenderer';
+
+import { FramesRenderer, type ThumbnailWithId } from './FramesRenderer';
+
+// Mock the SVG utilities
+jest.mock('teleport/SessionRecordings/svg', () => ({
+  generateTerminalSVGStyleTag: jest.fn(() => '<style></style>'),
+  injectSVGStyles: jest.fn(svg => svg),
+  svgToDataURIBase64: jest.fn(svg => `data:image/svg+xml;base64,${svg}`),
+}));
+
+beforeEach(() => {
+  Object.defineProperty(window, 'devicePixelRatio', {
+    writable: true,
+    configurable: true,
+    value: 1,
+  });
+
+  global.OffscreenCanvas = jest.fn().mockImplementation((width, height) => {
+    const canvas = document.createElement('canvas');
+
+    canvas.width = width;
+    canvas.height = height;
+
+    return canvas;
+  });
+});
+
+afterEach(() => {
+  global.OffscreenCanvas = undefined;
+});
+
+function createThumbnail(
+  svg: string,
+  cols: number,
+  rows: number,
+  startOffset: number,
+  cursorX = 0,
+  cursorY = 0
+): SessionRecordingThumbnail {
+  return {
+    svg,
+    startOffset,
+    endOffset: startOffset + 1000,
+    cols,
+    rows,
+    cursorX,
+    cursorY,
+    cursorVisible: true,
+  };
+}
+
+const mockFrames: SessionRecordingThumbnail[] = [
+  createThumbnail('<svg>frame1</svg>', 80, 24, 0),
+  createThumbnail('<svg>frame2</svg>', 100, 30, 1000),
+  createThumbnail('<svg>frame3</svg>', 120, 40, 2000),
+  createThumbnail('<svg>frame4</svg>', 80, 24, 3000),
+];
+
+async function createRenderer(
+  frames: SessionRecordingThumbnail[] = mockFrames,
+  timelineWidth = 1000,
+  duration = 5000,
+  initialHeight = 200,
+  eventsHeight = 50,
+  frameWidth = 200,
+  frameHeight = 50
+) {
+  const canvas = document.createElement('canvas');
+  const ctx = canvas.getContext('2d')!;
+
+  const renderer = new FramesRenderer(
+    ctx,
+    darkTheme,
+    duration,
+    frames,
+    initialHeight,
+    eventsHeight
+  );
+
+  renderer.setTimelineWidth(timelineWidth);
+  renderer.calculate();
+
+  jest
+    .spyOn(
+      // hack to be able to spy on a private method
+      // jest cannot load images in tests so we mock this method
+      renderer as unknown as {
+        loadImage: (frame: ThumbnailWithId) => Promise<OffscreenCanvas>;
+      },
+      'loadImage'
```

Comment:
Would passing `imageLoader` to `FramesRenderer` constructor help? 
```ts
 constructor(
    ctx: CanvasRenderingContext2D,
    theme: DefaultTheme,
    duration: number,
    frames: SessionRecordingThumbnail[],
    initialHeight: number,
    eventsHeight: number,
    private imageLoader: (
      frame: ThumbnailWithId
    ) => Promise<OffscreenCanvas> = loadImage
  )
```
We could have a custom implementation here.

Reply:
Yep great solution